### DISCUSSION TITLE
Query products by metafield value not working/filtering

### DISCUSSION THREAD

**Discussion by charle-connoringold** [Shopify/storefront-api-feedback#139] [Help] [answered]
Hi Everyone.

I'm new to Graphql, so I could be making a mistake, but here in my screenshot, you can see the filter isn't being correctly applied 😢

![image](https://user-images.githubusercontent.com/84711365/193019984-a01bd2ff-e2d1-4eda-9103-5656d059c97f.png)

Here is the code too:

```
query Diamonds {
  collection(handle: "best-selling-diamonds") {
    handle
    products(first: 3, 
      filters: {
        variantMetafield: {
          namespace:"Diamond",
          key:"Shape",
          value:"Oval"
      }
    }) {
      edges {
        node {
          handle
          title
          metafield: metafield(namespace: "Diamond", key:"Shape"){
            value
            description
            type
          }
        }
      }
    }
  }
}
```

I also have enabled all the meta fields in the Shopify Search & Discovery App:
![image](https://user-images.githubusercontent.com/84711365/193019652-24068fd0-cbee-47f7-a8c6-6adad836f9b4.png)

Any help is appreciated.

**Comment by johncraigcole** [comment, marked as answer]
Hey @charle-connoringold couple of things to review.

- Check that for the variant metafield filter the case of the `namespace` and `key` in the filter are correct. If these metafields were created with a definition in the admin or API they should be lower case. If the case is wrong it will be as though the filter was not applied.
- In your results you're highlighting a product metafield. Do you have a similar metafield defined at the product and variant level? Since you're filtering by variant metafield, seeing what is displayed for the variant metafield in the results would be worth checking as well.

**Comment by charle-connoringold** [comment]
Hey @johncraigcole 

So I realised after reading your comment I made some silly mistakes. As you mentioned, I was calling variant meta fields when I should have been calling product meta fields.

The second thing I did was redo all the meta fields to use lowercase namespaces. I wasn't aware this was a problem and can't see it mentions anything about this on the Shopify docs, so would be good if it was on there.

Regardless that fixed my problem, so thank you very much 👍.

**Comment by dnpg** [comment]
@johncraigcole @charle-connoringold I'm actually having this issue today and I'm actually using productMetafield as the filter

In my case I'm expecting to see only up to 5 results of product that contain the value "12 Colour Mountains", but instead I'm getting the default products listing without filtering. 

Note: Using api version 2022-10
The type of the metafield is single_line_text_field as specified in the docs, and can see in the screenshot attached.

Am I missing anything, an insight would be helpful, kind regards!


![Screen Shot 2022-10-21 at 2 36 01 pm](https://user-images.githubusercontent.com/1388364/197105942-2dc4b9c0-5009-4118-8911-7da74b980a6b.png)

**Reply by dnpg** [reply]
Nevermind, I found out that the merchant had disabled the filter in the Search & Discovery app. After enabling it the query worked as expected.

It would be great to see that we could filter here by any metafield without relying on the visible ones on the site. In my case it seems that the merchant doens't want to allow customer to filter out using that metafield anymore, but he'd like us to filter out some products in a custom section on the site.

**Comment by sk29110** [comment]
So metafield filtering doesn't work without search and discover app? I though I can just use graphql store front api to get this working.

**Reply by charle-connoringold** [reply]
That appears to be the case at the moment.

**Comment by sk29110** [comment]
Hello I got set up "this filter by metafield working with graphql" but when I am expecting 12 results I only manage to fetch 5 products . There are more than 1000 products with this collection. If I create another collection with smaller products like 50 then I manage to fetch all 12 results. Is this bug with store frontend api bug or the limitation?

```
{
	collection(id: "gid://shopify/Collection/1111111") {
		id
		handle
		title
		description
		products(
			first: 50
			filters: {
				productMetafield: {
					namespace: "pdp_fields"
					key: "group_handle"
					value: "somevalue"
				}
			}
		) {
			edges {
				node {
					id
					title
				}
			}
		}
	}
}
```

**Reply by johncraigcole** [reply]
Hey @sk29110 just wanted to confirm that the product metafield filter is specified as a filter in the Search & Discovery App in the admin. The behavior described indicates that it might not be. Assuming the filter has been added to the filters section of the Search & Discovery App it should work as you expect.

**Reply by sk29110** [reply]
@johncraigcole Thanks for reply. If I didn't specify the filter in S & D app I would get the all results back. It seems like there is the limitation on filtering the collection more than 1000 products. I decided to do with tag filtering instead so it works now. https://www.reddit.com/r/shopify/comments/tesb8d/help_collections_filters_dont_work_when_more_than/

**Reply by johncraigcole** [reply]
Hey @sk29110 👋 sorry for the late reply here. I've verified that yes the Storefront API is only sampling the first 1000 products of the given collection. So if the there's a collection of three thousand products, then some products that match the filter are are outside the first 1000 will not be returned on the result-set. There is currently work underway to fix this and not have any limit, but I don't have any timelines I can comfortably share with you today.

**Reply by johncraigcole** [reply]
Hey @sk29110 happy to share we've increased the product filtering in a collection to the first 5000 products, and we don't want to stop there. I don't have timelines on when an another increase my happen but hopefully the move from 1K to 5K is helpful.

**Comment by jokull** [comment]
Why isn't metafield filtering available on `products {}`, only `collections { products() ... }`?

**Reply by johncraigcole** [reply]
The initial implementation was to answer a common problem merchants both had with the online line store (liquid) and the Storefront API, and that was rich filtering support within the context of a collection. 

We are looking at supporting the same rich filtering of products outside the context of a collection as you'd like to see, but I can't share any timelines on when that might be available.

**Reply by hendrik244** [reply]
Hi @johncraigcole, do you have any updates on this?
Would be a very appreciated feature.

**Reply by johncraigcole** [reply]
Hey @hendrik244 we did introduce some new search capabilities to the Storefront API since I made that post. Specifically, [search](https://shopify.dev/docs/api/storefront/2023-07/queries/search) and [predictiveSearch](https://shopify.dev/docs/api/storefront/2023-07/queries/predictiveSearch). The `search` query supports product metafield filtering outside the context of a collection.

**Reply by hendrik244** [reply]
@johncraigcole Great, thank you.

**Comment by charle-connoringold** [comment]
@johncraigcole Do you know if/when we might be able to use operators on meta fields? 

For example, if we have a weight_in_kg metafield and want to find all products that have a metafield.weight_in_kg value great than 5.

Hopefully, that makes sense.

**Comment by SunboX** [comment]
Is there any way to filter customers like products based on metafield value?
